docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+## Code structure
+
+`Vector128<T>` is the "common denominator" across all platforms that support vectorization (and this is expected to always be the case). It represents a 128-bit vector of type `T`.
+
+`T` is constrained to specific primitive types:
+
+* `byte` and `sbyte` (8 bits).
+* `short` and `ushort` (16 bits).
+* `int`, `uint` and `float` (32 bits).
+* `long`, `ulong` and `double` (64 bits).
+* `nint` and `nuint` (32 or 64 bits, depending on the architecture, available in .NET 7+)
+
+.NET 8 introduced a `Vector128<T>.IsSupported` that indicates whether a given `T` will throw to help identify what works per runtime, including from generic contexts.
+
+A single `Vector128` operation allows you to operate on: 16 (s)bytes, 8 (u)shorts, 4 (u)ints/floats, or 2 (u)longs/double(s).
+
+```
+------------------------------128-bits---------------------------
+|             64                |               64              |
+-----------------------------------------------------------------
+|      32       |      32       |      32       |      32       |
+----------------------------------------------------------------|
+|  16   |  16   |  16   |  16   |  16   |  16   |  16   |  16   |
+-----------------------------------------------------------------
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough and the benchmarks prove that it offers better performance, you should use it instead of `Vector128<T>`. Namely, `Vector256<T>` on x86/x64 is mostly treated as `2x Vector128<T>` and while there are some operations that can "cross lanes", they can sometimes be more expensive or have other hidden costs.
+
+To check the acceleration, use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
+
+The size of the input also matters. It needs to be at least of the size of a single vector to be able to execute the vectorized code path (there are some advanced tricks that can allow you to operate on smaller inputs, but we won't describe them here). `Vector128<T>.Count` and `Vector256<T>.Count` return the number of elements of the given type T in a single vector.
+Both `Count` and `IsHardwareAccelerated` are turned into constants by the Just-In-Time compiler (i.e. no method call is required to retrieve the information). In the case of pre-compiled code (NativeAOT), this is not true for the `IsHardwareAccelerated` property, as the required information is not available at compile time.
+
+**Note:** When `Vector256` is accelerated then `Vector128` and `Vector64` are also accelerated.
```

Comment:
Today, when `Vector256` is accelerated than `Vector128` will also be; and we expect that will always be the case but isn't something we can strictly "guarantee".

Today, `Vector64` is not accelerated on x86/x64 (we could do some things here and make it work, but we don't today).

The reason I say we can't strictly guarantee it is there may be platforms where they only provide a variable sized vector and so they would only accelerate the largest size they provide.

Reply:
In general, the best practice for `IsHardwareAccelerated` is to always test explicitly. You can rely on certain behaviors for specific platforms (such as if `Vector256.IsHardwareAccelerated` on x86/x64, then `Vector128.IsHardwareAccelerated` will definitely be true), but not all of them.

The case where users really can rely on this is for the `Isa.IsSupported` checks as we guarantee that if `DerivedIsa.IsSupported` then `BaseIsa.IsSupported` must also be true. The same goes for if `Isa.NestedIsa.IsSupported` is true then `Isa.IsSupported` must be true.